Fix synchronization between MC6847 video display generator and screen device

dave-br opened the pull request:
@tlindner: FYI
 
- fix: #11567 

PROBLEM: The vertical position of the screen and the physical scanline of the MC6847 video display generator drift out of sync further and further every few frames.

PRIMARY FIX: Initialize the screen device using set_raw, with more precise parameters to describe horizontal and vertical blanking.  (Full list of supporting changes below.)

SCENARIOS FIXED:

1) mame 6809 debugging (my primary objective): Hitting F7 (run until next interrupt) stops at the next FS interrupt.  Before this fix, the debugger reports a beamy (based on the screen's vertical position) that varies all over the frame.  About every 3 frames (3 hits of F7), beamy goes up by one, through the entire range of vertical positions due to a continuous drift between the screen's and 6847's timing.  After this fix, beamy no longer drifts further and further away from the VDG's vertical position.  Beamy reliably rests in the range between the end of 216 and beginning of 217.

2) Dragonfire: Not my primary objective with this fix, but I tried out this game since its code is meticulously timed to allow multiple display mode changes per line.  It is still not perfect, but is significantly better.  The changing video garbage of the margins is gone, and the flashing on the sides of the castle walls is 99% gone.  There are still static fringes of purple around the moat.

3) When viewing logging, it is now clear that the 6847's scanline matches the screen's vpos.  The debugger scenario no longer drifts, but is imperfect due to varying cycles expended between an interrupt getting issued and the debugger stopping.  Logging, on the other hand, reliably shows the exact same screen hpos & vpos each time the 6847 advances to a new scanline or issues the FS timer.

OTHER EFFECTS:

1) The L/R margins were smaller than what a CoCo 1/2 normally generates.  I believe they are now the correct size.

2) Due to how the screen device handles the larger number of pixels I'm feeding it, the active display area now appears a bit more stretched vertically than prior builds.  This was jarring to me at first, but when I compare with my CoCo on an old CRT, it looks similar.

LIST OF CODE CHANGES:

Initialize screen device using raw parameters
Inform screen device of full vblanking and hblanking amounts
All hblanking moved from beginning of line to end of line (vblanking remains at end of field)
New logging on next scanline, more detailed logging elsewhere
Moved some repeatedly-used hard-coded numbers into #defines

angelosa commented on src/devices/video/mc6847.cpp:
```diff
@@ -76,25 +76,42 @@
-#define TIMER_FSYNC_OFF_TIME    (TIMER_HSYNC_PERIOD * TOP_BORDER + TIMER_HSYNC_ON_TIME)
-#define TIMER_FSYNC_ON_TIME     (TIMER_HSYNC_PERIOD * (TOP_BORDER + 192) + TIMER_HSYNC_ON_TIME)
 
 #define LOG_SCANLINE (1U << 1)
 #define LOG_HSYNC    (1U << 2)
 #define LOG_FSYNC    (1U << 3)
 #define LOG_FLUSH    (1U << 4)
 #define LOG_INPUT    (1U << 5)
+#define LOG_NEXTLINE (1U << 6)
 #define VERBOSE (0)
 #include "logmacro.h"
```
This block should really be just after the header declarations, as conventionally done by basically any device/driver that uses `logmacro.h`

dave-br replied:
Done.

angelosa commented on src/devices/video/mc6847.cpp:
```diff
@@ -406,6 +432,9 @@ inline void mc6847_friend_device::next_scanline()
 		enter_bottom_border();
 	}
+
+	// This helps to confirm synchronization between the screen device and the 6847
+	LOGMASKED(LOG_NEXTLINE, "mc6847_friend_device::next_scanline(): (vpos,hpos)=(%d,%d), m_physical_scanline='%d', m_logical_scanline_zone='%s', m_logical_scanline='%d'\n", (screen().vpos()), (screen().hpos()), m_physical_scanline, (scanline_zone_string((scanline_zone)m_logical_scanline_zone)), m_logical_scanline);
```
I feel like the comment is redunant, at most it should be on same line of the `LOG_NEXTLINE` definition itself above.

dave-br replied:
Makes sense.  I've removed the comment.

angelosa commented on src/devices/video/mc6847.cpp:
```diff
@@ -752,11 +793,36 @@ void mc6847_base_device::record_partial_body_scanline(uint16_t physical_scanline
 int32_t mc6847_base_device::scanline_position_from_clock(int32_t clocks_since_hsync)
 {
-	return (clocks_since_hsync - 20) / 4;
+	// This value determined by experimentation with the Dragon Fire game,
+	// which uses precise timing measurements to change video modes
+	// multiple times mid-line.  This value seems to minimize video
+	// garbage, but doesn't eliminate it, so there are likely timing
+	// issues elsewhere.
+	//
```
Dragon Fire is under `drgnfire` in `coco_cart.xml, please use the romset ID here. I otherwise give it a shot later, I presume it's main system waitstates?

dave-br replied:
Thanks for the info about the ROM.  I'd been using the disk version from here: https://colorcomputerarchive.com/repo/Disks/Games/Dragonfire%20(Tandy).zip (LOADM + EXEC).  I will look for the ROM and try it out next.  I haven't tried disassembling Dragonfire to see what it's doing, but debugging MAME itself (C++ debugger), I can see that Dragonfire uses the hsync ON interrupt.  My guess is that it responds to that interrupt by clearing a flag somewhere so that the mainline code is released at a known time to do its meticulously-timed processing for the next horizontal line.  Just a guess, though.

dave-br replied:
Tried the cartridge version from https://colorcomputerarchive.com/repo/Cartridges/Dragon%20Fire%20(1984)%20(26-3098)%20(Tandy).ccc.  Performs exactly the same as the disk version before my change (moving & shimmering garbage), and the same as the disk version after my change (nearly no garbage).  Will add "drgnfire" to the comment.

dave-br replied:
I've added "[Dragon Fire listed in coco_cart.xml as drgnfire.]" to the end of the comment.

angelosa replied:
That's even more redundant now, just referring it as `coco_cart:drgnfire` or `drgnfire (coco_cart)` is enough. 
It's a QA convention, adding Long Name Game Or Software Title as a comment is not pretty to look at, plus it will eventually become outdated the moment somebody takes this (and PC-6001) to a serious testing level.

dave-br replied:
I see.  Ok, I've replaced "dragon fire" with coco_cart:drgnfire.  There was also a prior comment at the top of the file that had referred to "dragon fire", so I replaced that as well.

startaq commented on src/devices/video/mc6847.cpp:
```diff
@@ -1,4 +1,4 @@
-// license:BSD-3-Clause
+// license:BSD-3-Clause 
```
You can avoid spurious changes like this by running `srcclean` on the file.

dave-br replied:
Nice, ran the tool this time.

angelosa commented on src/devices/video/mc6847.cpp:
```diff
@@ -559,11 +588,23 @@ void mc6847_base_device::device_config_complete()
+		// 256 pixels of active video area per line, with apparently no roundoff
+		// error or drift.
+		//
+		// Note: This math is valid for NTSC.  If / when PAL is implemented, m_tpfs
+		// can be used to determine if this 6847 is PAL, and a different set_raw
+		// call can be made if helpful.
```
I wonder what the real 6847 really do in order to pinpoint that is a PAL machine?
From quick glance from the datasheets the VDG seems coupled with a MC1372, which is what drives the clock on a NTSC machine, no idea about PAL atm but regardless I guess that `m_tpfs` is some kind of cheap simplification that should be removed in the end.

dave-br replied:
Yeah, I wish I had a good answer to that, but I'm brand new to all of this (hardware and MAME).  The original MC6847.cpp code switched off of m_tpfs when setting the refresh rate.  I wanted to preserve that approach in a comment when I changed the code to use set_raw, so that future changes will at least have a starting point to consider.

cuavas commented on src/devices/video/mc6847.cpp:
```diff
@@ -76,25 +76,42 @@
 
-#define TOP_BORDER              25
-#define USE_HORIZONTAL_CLIP     false
+#define LINES_TOP_BORDER				(25)
+#define LINES_ACTIVE_VIDEO				(192)
+#define LINES_BOTTOM_BORDER				(26)
+#define LINES_VERTICAL_RETRACE			(6)
+#define LINES_UNTIL_RETRACE				(LINES_TOP_BORDER + LINES_ACTIVE_VIDEO + LINES_BOTTOM_BORDER)
+#define LINES_UNTIL_VBLANK				(LINES_UNTIL_RETRACE + LINES_VERTICAL_RETRACE)
+
+#define USE_HORIZONTAL_CLIP				false
+
+#define CLOCKS_HSYNC_PERIOD				(228)
+#define	CLOCKS_FRONT_PORCH_DURATION		(7)
+#define CLOCKS_ACTIVE_VIDEO				(128)
+#define CLOCKS_L_OR_R_BORDER			(29)
+#define CLOCKS_ACTIVE_VIDEO_AND_BORDERS	(CLOCKS_ACTIVE_VIDEO + (2*CLOCKS_L_OR_R_BORDER))
+#define TIMER_HSYNC_OFF_TIME			(CLOCKS_ACTIVE_VIDEO_AND_BORDERS + CLOCKS_FRONT_PORCH_DURATION)
+#define TIMER_HSYNC_ON_TIME				(TIMER_HSYNC_OFF_TIME + 16.5)
+
+// FSYNC occurs just after active video, just before right border
+#define TIMER_FSYNC_TIME				(CLOCKS_ACTIVE_VIDEO + CLOCKS_L_OR_R_BORDER)
+
+// These units are the screen device's bitmap pixels.  Multiplied by 2 because
+// the pixel clock is the mc6847's clock * 2
+#define BMP_L_OR_R_BORDER				(CLOCKS_L_OR_R_BORDER * 2)
+#define BMP_ACTIVE_VIDEO				(CLOCKS_ACTIVE_VIDEO * 2)
```
Can you use `constexpr` in an anonymous namespace rather than macros?  It’s a lot nicer for compiler error messages and debugging symbols.

dave-br replied:
Agreed!  Done.

cuavas commented on src/devices/video/mc6847.cpp:
```diff
@@ -406,6 +432,9 @@ inline void mc6847_friend_device::next_scanline()
 		enter_bottom_border();
 	}
+
+	// This helps to confirm synchronization between the screen device and the 6847
+	LOGMASKED(LOG_NEXTLINE, "mc6847_friend_device::next_scanline(): (vpos,hpos)=(%d,%d), m_physical_scanline='%d', m_logical_scanline_zone='%s', m_logical_scanline='%d'\n", (screen().vpos()), (screen().hpos()), m_physical_scanline, (scanline_zone_string((scanline_zone)m_logical_scanline_zone)), m_logical_scanline);
```
Besides the comment being pointless, this line is ridiculously long and has redundant parentheses.

dave-br replied:
yup.  Comment is gone, and the LOGMASKED call has been broken into multiple lines.

cuavas commented on src/devices/video/mc6847.cpp:
```diff
@@ -559,11 +588,23 @@ void mc6847_base_device::device_config_complete()
+		// Note: This math is valid for NTSC.  If / when PAL is implemented, m_tpfs
+		// can be used to determine if this 6847 is PAL, and a different set_raw
+		// call can be made if helpful.
+		screen().set_raw(
+			(u32)(clock() * 2),
+			456,										// htotal
+			0,											// hbend
+			BMP_L_OR_R_BORDER * 2 + BMP_ACTIVE_VIDEO,	// hbstart
+			m_tpfs,										// vtotal
+			0,											// vbend
+			LINES_UNTIL_RETRACE);						// vbstart
```
The rest of this file doesn’t use the abbreviated integer types.  Please follow the pattern.

dave-br replied:
Indeed!  Changed.

cuavas commented on src/devices/video/mc6847.cpp:
```diff
@@ -161,9 +178,9 @@ void mc6847_friend_device::device_start()
-	m_frame_timer->adjust(clocks_to_attotime(0), 0, clocks_to_attotime(m_tpfs * TIMER_HSYNC_PERIOD * m_divider));
-	m_hsync_on_timer->adjust(clocks_to_attotime(TIMER_HSYNC_ON_TIME * m_divider), 1, clocks_to_attotime(TIMER_HSYNC_PERIOD * m_divider));
-	m_hsync_off_timer->adjust(clocks_to_attotime(TIMER_HSYNC_OFF_TIME * m_divider), 0, clocks_to_attotime(TIMER_HSYNC_PERIOD * m_divider));
+	m_frame_timer->adjust(clocks_to_attotime(0), 0, clocks_to_attotime(m_tpfs * CLOCKS_HSYNC_PERIOD * m_divider));
+	m_hsync_on_timer->adjust(clocks_to_attotime(TIMER_HSYNC_ON_TIME * m_divider), 1, clocks_to_attotime(CLOCKS_HSYNC_PERIOD * m_divider));
+	m_hsync_off_timer->adjust(clocks_to_attotime(TIMER_HSYNC_OFF_TIME * m_divider), 0, clocks_to_attotime(CLOCKS_HSYNC_PERIOD * m_divider));
```
`clocks_to_attotime` takes an integer parameter, but your `TIMER_HSYNC_ON_TIME` is a `double`.  Playing fast and loose with integer to floating point conversions leads to trouble.  It also won’t work with devices derived from `mc6847_base_device` which initialise `m_divider` as 1 as the 0.5 will get truncated away.

dave-br replied:
Nice catch.  m_divider is indeed 1 for the coco.  I think what I need is a clocks_to_attotime which multiplies its argument by m_attoseconds_per_clock using double arithmetic and then casts to an int.  I don't see anything like that--would you advise I create such a thing in device_t?  (Or is that generally too dangerous to legitimize at such a level?)

dave-br replied:
Actually, I'm changing 16.5 -> 16 altogether, and explicitly marking as int.  I tried the double arithmetic (multiplying TIMER_HSYNC_ON_TIME * m_attoseconds_per_clock) before explicitly casting, and that threw the timing off for Dragonfire.  I tried several adjustments, but interestingly can't do better than just straight int arithmetic with 16, which is what the code was inadvertently doing anyway.

In fact, it's not clear to me from the data sheet that 16.5 was the correct value to begin with.  There are small factors beyond 16.5 that affect HS width, including HS falling & rising time.  I'm likely missing something here, but 16 seems to be the most accurate number.

dave-br commented:
Thanks for all the feedback so far.  Please let me know if there is anything else you need from me.  I believe I've addressed all feedback so far, but just wanted to double-check you weren't waiting on me for anything.

angelosa commented:
I'm gonna apply this if/when CI passes, been forever I've been meaning to QA this system ...

dave-br commented:
Cool, thanks for the update.
